Libraries/LibWeb/Bindings/RangeConstructor.cpp

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2020, the SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibJS/Heap/Heap.h>
+#include <LibWeb/Bindings/RangeConstructor.h>
+#include <LibWeb/Bindings/RangeWrapper.h>
+#include <LibWeb/Bindings/WindowObject.h>
+#include <LibWeb/DOM/Range.h>
+
+namespace Web::Bindings {
+
+RangeConstructor::RangeConstructor(JS::GlobalObject& global_object)
+    : NativeFunction(*global_object.function_prototype())
+{
+}
+
+void RangeConstructor::initialize(JS::GlobalObject& global_object)
+{
+    NativeFunction::initialize(global_object);
+
+    define_property("length", JS::Value(0), JS::Attribute::Configurable);
+}
+
+JS::Value RangeConstructor::call()
+{
+    return construct(*this);
```

Comment:
I don't think `Range` is supposed to support construction without `new`.

```suggestion
    vm().throw_exception<JS::TypeError>(global_object(), JS::ErrorType::ConstructorWithoutNew, "Range");
    return {};
```

Reply:
I just took that from here:

https://github.com/SerenityOS/serenity/blob/48d2545572f1281bd9526f64798e51ecc862bb93/Libraries/LibWeb/Bindings/XMLHttpRequestConstructor.cpp#L60

If I understand you correctly, the JavaScript engine calls `RangeConstructor::construct` when the user writes `new Range` and calls `RangeConstructor::call` if the user calls it directly somehow?

Reply:
Apparently it can be constructed with `new`, but it can't be called as a function:
Chrome:
![chrome_BQeRmSXDsW](https://user-images.githubusercontent.com/25595356/101546824-42e8c100-39a1-11eb-9d59-30a5d457d31f.png)

Firefox:
![firefox_ujKusY64bR](https://user-images.githubusercontent.com/25595356/101546885-598f1800-39a1-11eb-8aad-b59bd30ead67.png)

Reply:
Precisely :)

- `Foo()` -> `::call()`
- `new Foo()` -> `::construct()`

In some cases the desired behaviour is them being equal, that's why some objects simply call `construct` in their `call`.

Looks like XHR shouldn't support construction without `new` either though :smile:

Reply:
Oh, that's what you mean, yes, it can be constructed with `new Range()` that just selects the entire document.

https://dom.spec.whatwg.org/#dom-range-range

Reply:
Now `Range()` throws an exception while `new Range()` works normally.